contracts/BonusRewards.sol

```diff
@@ -3,11 +3,369 @@
+        allowedTokenAuthorizers[_lpToken][_bonusTokenAddr] = _authorizers;
+        bonusTokenAddrMap[_bonusTokenAddr] = 1;
+      }
+    }
+  }
+
+  /// @notice collect bonus token dust to treasury
+  function collectDust(address _token, address _lpToken, uint256 _poolBonusId) external override onlyOwner {
+    require(pools[_token].lastUpdatedAt == 0, "BonusRewards: lpToken, not allowed");
+
+    if (_token == address(0)) { // token address(0) = ETH
+      payable(owner()).transfer(address(this).balance);
+    } else {
+      uint256 balance = IERC20(_token).balanceOf(address(this));
+      if (bonusTokenAddrMap[_token] == 1) {
+        // bonus token
+        Bonus memory bonus = pools[_lpToken].bonuses[_poolBonusId];
+        require(bonus.bonusTokenAddr == _token, "BonusRewards: wrong pool");
+        require(bonus.endTime + WEEK < block.timestamp, "BonusRewards: not ready");
+        balance = bonus.remBonus;
+        pools[_lpToken].bonuses[_poolBonusId].remBonus = 0;
+      }
+
+      IERC20(_token).transfer(owner(), balance);
+    }
+  }
+
+  function setResponders(address[] calldata _responders) external override onlyOwner {
+    responders = _responders;
+  }
+
+  function setPaused(bool _paused) external override {
+    require(_isAuthorized(msg.sender, responders), "BonusRewards: caller not responder");
+    paused = _paused;
+  }
+
+  function getPool(address _lpToken) external view override returns (Pool memory) {
+    return pools[_lpToken];
+  }
+
+  function getUser(address _lpToken, address _account) external view override returns (User memory, uint256[] memory) {
+    return (users[_lpToken][_account], viewRewards(_lpToken, _account));
+  }
+
+  function getAuthorizers(address _lpToken, address _bonusTokenAddr) external view override returns (address[] memory) {
+    return allowedTokenAuthorizers[_lpToken][_bonusTokenAddr];
+  }
+
+  function getResponders() external view override returns (address[] memory) {
+    return responders;
+  }
+
+  function viewRewards(address _lpToken, address _user) public view override returns (uint256[] memory) {
+    Pool memory pool = pools[_lpToken];
+    User memory user = users[_lpToken][_user];
+    uint256[] memory rewards = new uint256[](pool.bonuses.length);
+    if (user.amount <= 0) return rewards;
+
+    uint256 rewardsWriteoffsLen = user.rewardsWriteoffs.length;
+    for (uint256 i = 0; i < rewards.length; i ++) {
+      Bonus memory bonus = pool.bonuses[i];
+      if (bonus.startTime < block.timestamp && bonus.remBonus > 0) {
+        uint256 lpTotal = IERC20(_lpToken).balanceOf(address(this));
+        uint256 bonusForTime = _calRewardsForTime(bonus, pool.lastUpdatedAt);
+        uint256 bonusPerToken = bonus.accRewardsPerToken + bonusForTime / lpTotal;
+        uint256 rewardsWriteoff = rewardsWriteoffsLen <= i ? 0 : user.rewardsWriteoffs[i];
+        uint256 reward = user.amount * bonusPerToken / CAL_MULTIPLIER - rewardsWriteoff;
```

Comment:
Why subtract by rewards writeoff here?

Reply:
The way we calculate a user's rewards is by
`total rewards accumulated so far on each lpToken * user.amount - user write off on the amount so far` to fake an `integral D_amount D_time`.